### ISSUE TITLE
[Q] Was the data cleaned in any way prior to being processed? 

### ISSUE THREAD

**Issue by theace0296** [wvsharber/BeatMapSynthesizer#37]
You said that you only used mappings with user ratings above 90% and 70% for the two datasets, but did you also sort out the mappings that have a low number of ratings? If not, I would suggest finding a reasonable cut-off for the number of ratings required to eliminate any "high" ratings because the maps haven't been played or rated by that many people.

**Comment by wvsharber** [comment]
Hi! 

That's a great question. No, I didn't provide a cut-off for the number of ratings for each song. You're right though, that could potentially improve the quality of the mappings. The models I've trained here are a first pass. I'm sure some great maps have been uploaded to Beat Saver since I first downloaded the data in January. I would love to retrain the models with new data, and potentially take into account the number of ratings as well.

That being said, I think including as many different user-made mappings in the training data as possible is a good thing if you want the potential to have variety in the created maps. In my experience playing the top 90% (version 1) vs top 70% (version 2) models, the variety of moves increases in the 70% models because there a lot more potential moves in the Hidden Markov Models transition matrices. 

I also have some ideas about how to improve some of the modeling steps, so retraining the models is probably worth a revisit. I just need the time to work on it again! Right now I'm focusing on making the program more user friendly (developing a standalone executable) and doing scientific user studies of model performance. 

I hope that answers your question though! Let me know if you have anymore!

**Comment by theace0296** [comment]
Thanks for responding! I'm not all too familiar with writing GUI's in Python beyond some of the basics, but I did do some work on your program for doing batch processing and parallel processing. I wouldn't say what I've done is worth a pull request or anything, but from my testing on maybe a few hundred songs I ran through it, the biggest issues I ran into were very some notes getting very rapid awkward movements (an example being many down-strikes in a row, rather than alternating strikes), and an overall difficulty with vocal-focused songs. The later could potentially be solved with time as the model gains more options for dealing with complicated songs. 

I'm looking forward to seeing this progress!

**Comment by wvsharber** [comment]
The GUI I'm developing is definitely very basic, but hopefully it will be easier to use for most people. I made this program as an exercise in data modeling, not so much in software development.

I'd be interested in seeing what you wrote! If you have a separate .py file that you created or something, feel free to send it to me. My username is the same on discord and my gmail account.

As far as the issues that you've run into, those are things I've seen too. Songs without strong beats or with inconsistent beats struggle because of the way beats are identified, so timing can get off in those kinds of songs. The awkward rapid sequences seem especially problematic when playing on expert or expert plus. I think the best option would to have a separate model for very rapid block sequences (e.g., "drumrolls") that is used specifically during parts of a song that are appropriate for a rapid block sequence. That being said, there may be better ways of approaching the modeling process in general that could improve performance. I don't know if you've seen it, but there is a neural network mapper called Beat Sage that uses a two step network to first plan where in a song to place blocks and then secondly determines which blocks to place. BeatMapSynth is similar in the two-step approach, but the first step is rules-based with some randomness added and the second step is dependent on the Hidden Markov Models. I think the main advantage with the BeatMapSynth approach is that it's quick and can run on anyone's PC, rather than needing GPUs to run the neural networks in Beat Sage. There's always room for improvement though!

Let me know if you have anymore questions or comments! Thanks for taking the time to use it!

**Comment by theace0296** [comment]
The basics of what I did can be found in the following code. I had to make some tweaks to your original code and haven't completely changed up how some of the models are called, but I was mainly testing with the rate_modulated_segmented_HMM_mapper function. If I polish out the bugs I was dealing with, I'll add a pull request, but I would want to make it more user friendly first.

```
def getListOfFiles(dirName):
    # create a list of file and sub directories 
    # names in the given directory 
    listOfFile = os.listdir(dirName)
    allFiles = list()
    # Iterate over all the entries
    for entry in listOfFile:
        # Create full path
        fullPath = os.path.join(dirName, entry)
        # If entry is a directory then get the list of files in this directory 
        if os.path.isdir(fullPath):
            allFiles = allFiles + getListOfFiles(fullPath)
        else:
            allFiles.append(fullPath)
                
    return allFiles
    
def getListOfMusicFiles():
    dirName = 'A:\BeatMapSynthesizer\Music';
    
    # Get the list of all files in directory tree at given path
    listOfFiles = getListOfFiles(dirName)
    listOfMusicFiles = []
    
    # Print the files
    for file in listOfFiles:
        if file.endswith(".mp3"):
            listOfMusicFiles.append(file)
    return listOfMusicFiles
    
def generate_song_name_from_metadata(song_path):
    song_name = 'default'
    audiofile = eyed3.load(song_path)
    artist = audiofile.tag.artist
    track = audiofile.tag.title
    if artist is not None and track is not None:
        song_name = track + ' - ' + artist
    elif track is not None:
        song_name = track
    elif artist is not None:
        song_name = artist   
    
    invalidchars = [ "<", ">", ":", '"', "/", "\\", "|", "?", "*" ]
    if any(char in song_name for char in invalidchars):
        for char in invalidchars:
            song_name.replace(char, "^")
            
    return song_name
    
def run_model(song_path, song_name, difficulty, model, k=5, version=2):
    if model == 'random':
        random_mapper(song_path, song_name, difficulty)
    elif model == 'HMM':
        HMM_mapper(song_path, song_name, difficulty, version = version)
    elif model == 'segmented_HMM':
        segmented_HMM_mapper(song_path, song_name, difficulty, k = k, version = version)
    elif model == 'rate_modulated_segmented_HMM':
        rate_modulated_segmented_HMM_mapper(song_path, song_name, difficulty, version = version, k = k)
    else:
        print('Please specify model for mapping.')

def batch_beat_map(song_path, song_name, difficulty, version, k):
    
    song_name = generate_song_name_from_metadata(song_path)
                
    if not os.path.exists(f"./generated_tracks/{song_name}"):
        os.makedirs(f"./generated_tracks/{song_name}")
        try:
            run_model(song_path, song_name, difficulty, version = version, k = k)
        except:
            print(f"Error occured during: {difficulty} | {song_name}")
    elif not os.path.exists(f"./generated_tracks/{song_name}/normal.dat"):
        try:
            run_model(song_path, song_name, 'normal', version = version, k = k)
        except:
            print(f"Error occured during: normal | {song_name}")
    elif not os.path.exists(f"./generated_tracks/{song_name}/hard.dat"):
        try:
            run_model(song_path, song_name, 'hard', version = version, k = k)
        except:
            print(f"Error occured during: hard | {song_name}")
    elif not os.path.exists(f"./generated_tracks/{song_name}/expert.dat"):
        try:
            run_model(song_path, song_name, 'expert', version = version, k = k)
        except:
            print(f"Error occured during: expert | {song_name}")
    elif not os.path.exists(f"./generated_tracks/{song_name}/expertPlus.dat"):
        try:
            run_model(song_path, song_name, 'expertPlus', version = version, k = k)
        except:
            print(f"Error occured during: expertPlus | {song_name}")
    else:
        try:
            run_model(song_path, song_name, 'UPDATE', version = version, k = k)
        except:
            print(f"Error occured during: UPDATE | {song_name}")

#Main Function:
def beat_map_synthesizer(song_path, song_name, difficulty, model, k=8, version=2):
    """
    Function to load a music file and generate a custom Beat Saber map based on the specified model and difficulty. Outputs a zipped folder of necessary files to play the custom map in the Beat Saber game.
    
    ***
    song_path = string file path to music file location
    
    song_name = string to name level as it will appear in the game
    
    difficulty = desired difficulty level, can be: 'easy', 'normal', 'hard', 'expert', or 'expertPlus'
    
    model = desired model to use for map generation, can be: 'random', 'HMM', 'segmented_HMM', or 'rate_modulated_segmented_HMM'
    
    k = number of song segments if using a segmented model. Default is 5, may want to increase or decrease based on song complexity
   
    version = for HMM models, can choose either 1 or 2. 1 was trained on a smaller, but potentially higher quality dataset (custom maps with over 90% rating on beatsaver.com), while 2 was trained on a larger dataset of custom maps with over 70% rating, so it may have a larger pool of "potential moves."
    ***
    """
    # cd /d A:\BeatMapSynthesizer
    # conda activate BSMapSynth-env
    # python beatmapsynth.py "A:\BeatMapSynthesizer\Music" "ALL_SONGS" "ALL" "rate_modulated_segmented_HMM"
    
    if not os.path.exists('./generated_tracks/'):
        os.makedirs('./generated_tracks/')
    
    if (song_name == 'ALL_SONGS'):
        num_cores = multiprocessing.cpu_count() - 2
        musicList = getListOfMusicFiles()
        # During the parallel task, some errors cause all of the parallel tasks to quit
        # TODO: Implement a form of progress IPC and task restarting
        # This for loop is extremely inefficient and crude, but it does guarantee *almost* every song in the list gets run
        for n in range(0,len(musicList)):
            try:
                musicList = musicList[n:len(musicList)]
                Parallel(n_jobs=num_cores, timeout=300, verbose=0)(delayed(batch_beat_map)(file, song_name, difficulty, version, k) for file in musicList)
            except:
                continue
    else:
        run_model(song_path, song_name, difficulty, version = version, k = k)
```
